### PR TITLE
Detect workspace root using language markers

### PR DESCRIPTION
**Pull request by amousset**
Implement workspace root detection based on https://github.com/helix-editor/helix/issues/256#issuecomment-1001039937

### CODE REVIEW

File: helix-core/src/lib.rs [hunk 1]
```diff
@@ -39,8 +39,14 @@ pub fn find_first_non_whitespace_char(line: RopeSlice) -> Option<usize> {
     line.chars().position(|ch| !ch.is_whitespace())
 }
 
-/// Find `.git` root.
-pub fn find_root(root: Option<&str>) -> Option<std::path::PathBuf> {
+/// Find project root.
+///
+/// Order of detection:
+/// * Top-most folder containing a root marker in current git repository
+/// * Git repostory root if no marker detected
```
Comment (matoous) [inline helix-core/src/lib.rs:46]:
```suggestion
/// * Git repository root if no marker detected
```

File: helix-core/src/lib.rs [hunk 2]
```diff
@@ -39,8 +39,14 @@ pub fn find_first_non_whitespace_char(line: RopeSlice) -> Option<usize> {
...
+/// * Top-most folder containing a root marker if not git repository detected
+/// * Current working directory as fallback
+pub fn find_root(root: Option<&str>, root_markers: Vec<String>) -> Option<std::path::PathBuf> {
```
Comment (kirawi) [inline helix-core/src/lib.rs:49]: I think you can use `root_markers: &[String]` here.

File: helix-lsp/src/client.rs [hunk 3]
```diff
@@ -39,6 +40,7 @@ impl Client {
         cmd: &str,
         args: &[String],
         config: Option<Value>,
+        root_markers: Vec<String>,
```
Comment (kirawi) [inline helix-lsp/src/client.rs:43]: I think you should also be able to use `&[String]` here.

Comment (kirawi) [reply helix-lsp/src/client.rs:43]: Nevermind, you can resolve this.

File: helix-core/src/lib.rs [hunk 4]
```diff
@@ -52,16 +58,34 @@ pub fn find_root(root: Option<&str>) -> Option<std::path::PathBuf> {
...
+        }
+        // don't go higher than repo
         if ancestor.join(".git").is_dir() {
-            return Some(ancestor.to_path_buf());
+            // Use workspace if detected from marker
+            if top_marker.is_some() {
+                return top_marker.map(|a| a.to_path_buf());
+            } else {
+                return Some(ancestor.to_path_buf());
+            }
```
Comment (archseer) [inline helix-core/src/lib.rs:79]:
```suggestion
            return top_marker.unwrap_or(ancestor).to_path_buf();
```

File: helix-core/src/lib.rs [hunk 5]
```diff
@@ -52,16 +58,30 @@ pub fn find_root(root: Option<&str>) -> Option<std::path::PathBuf> {
...
+            if ancestor.join(marker).exists() {
+                top_marker = Some(ancestor);
+                break;
+            }
+        }
+        // don't go higher than repo
         if ancestor.join(".git").is_dir() {
-            return Some(ancestor.to_path_buf());
+            // Use workspace if detected from marker
+            return top_marker.unwrap_or(ancestor).to_path_buf();
```
Comment (kirawi) [inline helix-core/src/lib.rs:75]:
```suggestion
            return Some(top_marker.unwrap_or(ancestor).to_path_buf());
```

**Review by archseer — APPROVED** [review]
Thanks! 🎉

**Pull request merged**